server/src/main/java/io/druid/curator/discovery/CuratorDruidLeaderSelector.java

```diff
@@ -0,0 +1,204 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.curator.discovery;
+
+import com.google.common.base.Throwables;
+import com.metamx.emitter.EmittingLogger;
+import io.druid.concurrent.Execs;
+import io.druid.concurrent.LifecycleLock;
+import io.druid.discovery.DruidLeaderSelector;
+import io.druid.guice.annotations.Self;
+import io.druid.java.util.common.ISE;
+import io.druid.java.util.common.StringUtils;
+import io.druid.java.util.common.guava.CloseQuietly;
+import io.druid.server.DruidNode;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.recipes.leader.LeaderLatch;
+import org.apache.curator.framework.recipes.leader.LeaderLatchListener;
+import org.apache.curator.framework.recipes.leader.Participant;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ */
+public class CuratorDruidLeaderSelector implements DruidLeaderSelector
+{
+  private static final EmittingLogger log = new EmittingLogger(CuratorDruidLeaderSelector.class);
+
+  private final LifecycleLock lifecycleLock = new LifecycleLock();
+
+  private final DruidNode self;
+  private final CuratorFramework curator;
+  private final String latchPath;
+
+  private DruidLeaderSelector.Listener listener = null;
+  private final AtomicReference<LeaderLatch> leaderLatch = new AtomicReference<>();
+
+  private volatile boolean leader = false;
+  private volatile int term = 0;
+
+  public CuratorDruidLeaderSelector(CuratorFramework curator, @Self DruidNode self, String latchPath)
+  {
+    this.curator = curator;
+    this.self = self;
+    this.latchPath = latchPath;
+  }
+
+  private LeaderLatch createNewLeaderLatch()
+  {
+    final LeaderLatch newLeaderLatch = new LeaderLatch(
+        curator, latchPath, self.getHostAndPortToUse()
+    );
+
+    newLeaderLatch.addListener(
+        new LeaderLatchListener()
+        {
+          @Override
+          public void isLeader()
+          {
+            try {
+              if (leader) {
+                log.warn("I'm being asked to become leader. But I am already the leader. Ignored event.");
+                return;
+              }
+
+              leader = true;
+              term++;
+              listener.becomeLeader();
+            }
+            catch (Exception ex) {
+              log.makeAlert(ex, "listener becomeLeader() failed. Unable to become leader").emit();
+
+              // give others a chance to become leader.
+              final LeaderLatch oldLatch = createNewLeaderLatch();
+              CloseQuietly.close(oldLatch);
+              leader = false;
+              try {
+                //Small delay before starting the latch so that others waiting are chosen to become leader.
+                Thread.sleep(1000);
```

Comment:
An unlucky GC could easily induce a pause of 1 second. Is there a way to do this without the mandatory pause?

Reply:
I can't think of a better way right now without introducing a lot of potentially unhelpful complexity.

Reply:
https://github.com/druid-io/druid/issues/3428 is solved by this?

Reply:
overlord was using curator LeaderSelector and not LeaderLatch , which is a completely different implementation. Most likely #3428 would get solved by this.

Reply:
For stuff like this, a random sleep works better than a fixed sleep, since a fixed sleep can cause the same sequence of leader changes to play out over and over again. And I think a random sleep is a totally valid mechanism of trying to jolt a leader-election system onto a better leader. It's pretty common in leader-election algorithms.

In this case, maybe a random sleep between 500ms and 5000ms would work.

Reply:
makes sense, changed random sleep between 1000ms to 5000ms . 
min is 1000ms and not 500ms just to be safer and also because I tested with that value mostly.